nemo/collections/asr/models/ctc_models.py

```diff
@@ -655,18 +656,19 @@ def test_dataloader(self):
     def _transcribe_on_begin(self, audio, trcfg: TranscribeConfig):
         super()._transcribe_on_begin(audio, trcfg)
 
-        # Freeze the encoder and decoure_exder modules
-        self.encoder.freeze()
-        self.decoder.freeze()
+        # Freeze the encoder and decoder modules
```

Comment:
Can you revert all the removed sections that call freeze and unfreeze / as_frozen please ? 

We had a memory leak there for RNNT, it's fixed via inference_mode() but more often than note users do not remember to use it and face the memory leak. It is better for the freeze and unfreeze logic to be inherent on call so that even by mistake such leak is not possible

Reply:
That is precisely the problem. If you don't want to use torch.inference_mode() or torch.no_grad(), but instead rely on setting the requires_grad field of a tensor to false, then AMP will *always* fail to use its "casted tensor cache". So it will repeatedly cast your float32 weights to float16, which is what makes inference slow for all of our models.  It is because of the args.requires_grad() condition in this code: https://github.com/pytorch/pytorch/blob/6f5f405b057c7de0f5fce0b1432cb74468f96f95/aten/src/ATen/autocast_mode.cpp#L121C61-L121C80

The code is a rough draft for a reason. This is one way to fix the problem, but I don't think it's right because it is so fragile. I think that simply casting the model to float16 or bfloat16 would be more appropriate.

Reply:
How about we avoid mixed precision then and require full fp16 or bf16 cast (apart from all BN uses, which need to be wrapped to always cast to 32 no matter what, similar to how it's done in metas ASR library)

Reply:
I think that is the right solution. I'm frankly surprised NeMo uses AMP for inference, because it increases weight memory usage and isn't intended to be used for inference (and no other toolkits use it for inference either).

Of course... a lot of code is written around the assumption of AMP being used for inference. Take a look at the do_amp flag in in transcribe_speech.py, etc. So it's a heavy lift.